akka-stream/src/main/scala/akka/stream/impl/Timers.scala

```diff
@@ -140,63 +184,74 @@ private[stream] object Timers {
         override def onDownstreamFinish(): Unit = cancel(in2)
       })
 
-      private def onActivity(): Unit = nextDeadline = Deadline.now + timeout
+      private def onActivity(): Unit = nextDeadline = System.nanoTime + timeout.toNanos
 
       final override def onTimer(key: Any): Unit =
-        if (nextDeadline.isOverdue())
+        if (nextDeadline - System.nanoTime < 0)
           failStage(new TimeoutException(s"No elements passed in the last $timeout."))
 
-      override def preStart(): Unit = schedulePeriodically("IdleTimeoutCheckTimer", idleTimeoutCheckInterval(timeout))
+      override def preStart(): Unit = schedulePeriodically(IdleTimeoutBidiCheckTimer, idleTimeoutCheckInterval(timeout))
     }
+
+    override def toString = "IdleTimeoutBidi"
+
+    case object IdleTimeoutBidiCheckTimer
```

Comment:
One allocation per instance of IdleTimeoutBidi when it could be one shared instance (outside of the `IdleTimeoutBidi` class).